compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -100,12 +100,35 @@ object Applications {
       Nil
     }
 
+    def unapplySeqTypeElemTp(getTp: Type): Type = {
+      val lengthTp = ExprType(defn.IntType)
+      val lengthCompareTp = MethodType(List("len".toTermName))(_ => defn.IntType :: Nil, _ => defn.IntType)
+      def applyTp(elemTp: Type) = MethodType(List("i".toTermName))(_ => defn.IntType :: Nil, _ => elemTp)
+      def dropTp(elemTp: Type) = MethodType(List("n".toTermName))(_ => defn.IntType :: Nil, _ => defn.SeqType.appliedTo(elemTp))
+      def toSeqTp(elemTp: Type) = ExprType(defn.SeqType.appliedTo(elemTp))
+
+      val elemTp = getTp.member(nme.apply).suchThat(_.info <:< applyTp(WildcardType)).info.resultType
+
+      def names1 = List(nme.lengthCompare, nme.apply, nme.drop, nme.toSeq)
+      def types1 = List(lengthCompareTp, applyTp(elemTp), dropTp(elemTp), toSeqTp(elemTp))
+
+      def names2 = List(nme.length, nme.apply, nme.drop, nme.toSeq)
+      def types2 = List(lengthTp, applyTp(elemTp), dropTp(elemTp), toSeqTp(elemTp))
+
+      val valid = elemTp.exists &&
+                  (getTp <:< RefinedType.make(defn.AnyType, names1, types1) ||
```

Comment:
This seems like a rather complicated (and probably inefficient) way of checking if a type has a given member, why not use `extractorMemberType` like in other places in this file?

Reply:
I agree with @smarter. It seems inefficient. In particular the second subtype test `getTp <:< RefinedType.make(defn.AnyType, names2, types2))` will do a lot of unnecessary work that has already been done in the previous subtype test (i.e. checking members `apply`, `drop` and `toSeq`).

I think the right ways is to perform members lookup even if it is more verbose